[AMD-AIE] Add amenities for setting and using TranslationInfo and LoweringConfig

Abhishek-Varma opened the pull request:
This PR consists of two commits :-

**Commit 1 (Create TranslationInfo and LoweringConfig):**
-- This commit adds tile sizes and pass pipelines via lowering_config
   and translation_info respectively.
-- This is the initial effort towards generalizing the AIE codegen
   pipeline. Currently setting tile sizes suiting the Matmul pad based
   approach.

Signed-off-by: Nirvedh Meshram <nmeshram@amd.com>

**Commit 2 (Use TranslationInfo and LoweringConfig):**
-- This commit adapts the codegen flow to use lowering_config and
   translation_info.
-- This basically involves :-
   a. Adapting `--iree-amdaie-tile-and-fuse` to use lowering_config
      in order to fetch the tile size for a particular tiling level.
   b. Segregate the AMDAIE codegen flow to have two separate pipelines:
      i. Transform dialect pipeline ii. Multi level tiling C++ pipeline.

Signed-off-by: Abhishek Varma <abhvarma@amd.com>

Abhishek-Varma commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerExecutableTarget.cpp:
```diff
@@ -93,11 +140,24 @@ void AMDAIELowerExecutableTargetPass::runOnOperation() {
         addTransformDialectPasses(executableLoweringPipeline);
         break;
+      }
+      // TODO(avarma): Currently we are using "CPUDoubleTilingExpert".
+      //               Should we have "AIEDoubleTilingExpert" instead? This
+      //               seems to either be addressed in IREE repo or somehow we
+      //               are to extend "DispatchLoweringPassPipelineEnum" defined
+      //               inside IREECodegenAttrs.td.
+      case IREE::Codegen::DispatchLoweringPassPipeline::CPUDoubleTilingExpert: {
+        TilingConfig tilingConfig = getTilingConfigForPipeline(moduleOp);
+        addMultiTilingExpertPassPipeline(executableLoweringPipeline,
+                                         tilingConfig);
+        break;
+      }
```
Extending on the above CPU vs AIE vs Common discussion.

What should be our modus operandi here?

To me, it seems like we need to add the corresponding enums in IREECodegenAttrs.td.
Besides the general `CPU` -> `AIE` renaming, I believe this would be required to also tackle the PAD vs PACK pipeline approach in AIE.

makslevental commented:
I like the naming here - "amenities" - like you're getting the deluxe experience when you use IREE :smile:

Abhishek-Varma commented on tests/samples/matmul_fill_static_i32_config.mlir:
```diff
@@ -0,0 +1,39 @@
+//       CHECK:      aiex.ipu.dma_memcpy_nd
+//       CHECK:      aiex.ipu.sync
+#config = #iree_codegen.lowering_config<tile_sizes = [[8, 8], [4, 4], [0, 0, 4]]>
+#translation = #iree_codegen.translation_info<CPUDefault>
```
This is just to demonstrate that `TransformDialectCodegen` is not being set and the e2e still works.

MaheshRavishankar replied:
Just add this as a comment to the test itself.

Abhishek-Varma commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerExecutableTarget.cpp:
```diff
@@ -24,6 +25,12 @@
 #include "mlir/Pass/PassManager.h"
 #include "mlir/Pass/PassRegistry.h"
 #include "mlir/Transforms/Passes.h"
+// TODO(avarma):
+//     We shouldn't add "CPUUtils" - instead it should perhaps just be "Utils"?
```
I'll raise an upstream PR and address this comment as a follow-up when it gets merged in IREE. Marking this resolved for now.

MaheshRavishankar commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/KernelDispatch.cpp:
```diff
@@ -0,0 +1,116 @@
+/// Sets the lowering configuration for dispatch region with root op that
+/// implements the contraction operation interface.
+static LogicalResult setRootConfig(
+    func::FuncOp entryPointFn, linalg::ContractionOpInterface contractionOp) {
```
Just make this `linalg::MatmulOp` for now. We can generalize it once we try to expand coverage.

MaheshRavishankar commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.h:
```diff
@@ -25,6 +26,9 @@ void addTransformDialectPasses(OpPassManager &passManager);
 /// within the IREE::HAL::ExecutableOp.
 void buildAMDAIETransformPassPipeline(OpPassManager &pm);
 
+void addPadBasedPassPipeline(OpPassManager &passManager,
```
Add comments please.

MaheshRavishankar reviewed (approved):
Left a couple of more comments, but looks good to me. THanks!

Abhishek-Varma commented:
I have addressed the review comments and raised a PR for [CPUUtils -> Utils](https://github.com/openxla/iree/pull/16144) in upstream IREE.

Getting this in nonetheless.
